clang/lib/CConv/ConstraintBuilder.cpp

```diff
@@ -105,6 +109,20 @@ class FunctionVisitor : public RecursiveASTVisitor<FunctionVisitor> {
     return true;
   }
 
+  // Cast expressions must be visited to find generic functions where the return
+  // can be given a concrete type.
+  bool VisitCastExpr(CastExpr *CE){
```

Comment:
I'm not following what's happening here. Why (again) is visiting the `CallExpr` itself not sufficient? Also, aren't we already visiting cast expressions in `ConstraintBuilder` ? What happened to that code?

Reply:
Followup: Would it make sense to move this type param stuff into its own file? Is it a separate visitor?

Reply:
The type parameter information is used during constraint generation to decide if we can safely ignore implicit casts in functions calls, so we can't move it  to the rewriting stage. I might able to move it to a separate visitor that is executed before constraint generation.

Reply:
I like the idea of making it a separate visitor.

Reply:
I'll make those changes on top of the newer PR.

Reply:
OK